hosting/opennebula_functions.py

```diff
@@ -0,0 +1,230 @@
+                                              <DISK>
+                                                <TYPE>{disk_type}</TYPE>
+                                                <SIZE>{size}</SIZE>
+                                              </DISK>
+                                            </VM>
+                                            """
+                    vm_id = oca.VirtualMachine.allocate(self.client,
+                                                        vm_string_formatter.format(
+                                                            memory=1024 * vm_template_int,
+                                                            vcpu=vm_template_int,
+                                                            cpu=0.1 * vm_template_int,
+                                                            disk_type='fs',
+                                                            size=10000 * vm_template_int))
+                    message = _("Created with id = " + str(vm_id))
+                    messages.add_message(request, messages.SUCCESS, message)
+                else:
+                    messages.add_message(request, messages.ERROR,
+                                         _("Please select an appropriate value for vm template."))
+            except socket.timeout as socket_err:
+                messages.add_message(request, messages.ERROR, _("Socket timeout error."))
+                logger.error("Socket timeout error: {0}".format(socket_err))
+            except OpenNebulaException as opennebula_err:
+                messages.add_message(request, messages.ERROR,
+                                     _("OpenNebulaException occurred. {0}".format(opennebula_err)))
+                logger.error("OpenNebulaException error: {0}".format(opennebula_err))
+            except OSError as os_err:
+                messages.add_message(request, messages.ERROR, _("OS error: {0}".format(os_err)))
+                logger.error("OSError : {0}".format(os_err))
+            except ValueError as value_err:
+                messages.add_message(request, messages.ERROR,
+                                     _("Please select an appropriate value for vm template."))
+                logger.error("ValueError : {0}".format(value_err))
+        return redirect('admin:showvms')
+
+    # Delete VM from the pool and DB by using method finalize()
+    def delete_vm(self, request, vmid):
+        vm_id = int(vmid)
+        # get the desired vm from the pool
+        logger.debug("Deleting vm with id {0}".format(vm_id))
+        vm = self.get_vm_by_id(vm_id)
+        if vm is None:
+            messages.add_message(request, messages.ERROR, _("Did not find a vm with id = {0}".format(vm_id)))
+        else:
+            logger.debug("Deleting vm_id = " + str(vm_id) + " state = " + vm.str_state)
+            if vm.str_state == 'PENDING' or vm.str_state == 'POWEROFF' or vm.str_state == 'ACTIVE':
```

Comment:
If we are keep using the str_state here, can we at lease use constants for these values?